components/sup/src/manager/spec_watcher.rs

```diff
@@ -0,0 +1,600 @@
+        // we should attempt to update a service in-place, if possible.
+        for name in active_names.intersection(&desired_names) {
+            let active_spec = active_specs.remove(name).expect("value should exist for key");
+            let desired_spec = desired_specs.remove(name).expect("value should exist for key");
+            if active_spec != desired_spec {
+                let remove_event = SpecWatcherEvent::RemoveService(active_spec);
+                let add_event = SpecWatcherEvent::AddService(desired_spec);
+                debug!("Service spec for {} is different on disk than loaded state, \
+                       enqueuing {:?} for existing and {:?} event for updated spec",
+                       &name,
+                       &remove_event,
+                       &add_event);
+                events.push(remove_event);
+                events.push(add_event);
+            }
+        }
+
+        // Both maps should be empty, meaning we've processed them all
+        assert!(active_specs.is_empty());
+        assert!(desired_specs.is_empty());
+
+        Ok(events)
+    }
+
+    fn specs_from_watch_path<'a>(&self) -> Result<HashMap<String, ServiceSpec>> {
+        let spec_files: Vec<PathBuf> = glob(&self.watch_path
+                                                 .join(SPEC_FILE_GLOB)
+                                                 .display()
+                                                 .to_string())
+                ?
+                .filter_map(|p| p.ok())
+                .filter(|p| p.is_file())
+                .collect();
+
+        let mut specs = HashMap::new();
+        for spec_file in spec_files {
+            let spec = match ServiceSpec::from_file(&spec_file) {
+                Ok(s) => s,
+                Err(e) => {
+                    match e.err {
+                        // If the error is related to loading a `ServiceSpec`, emit a warning
+                        // message and continue on to the next spec file. The best we can do to
+                        // fail-safe is report and skip.
+                        Error::ServiceSpecFileRead(_, _) |
+                        Error::ServiceSpecParse(_) |
+                        Error::MissingRequiredIdent => {
+                            outputln!("Error when loading service spec file '{}' ({}). \
+                                      This file will be skipped.",
+                                      spec_file.display(),
+                                      e.description());
+                            continue;
+                        }
+                        // All other errors are unexpected and should be dealt with up the calling
+                        // stack.
+                        _ => return Err(e),
+                    }
+                }
+            };
+            let file_stem = spec_file.file_name()
+                .expect("name should not be ..")
```

Comment:
It looks like these `expect()`s could be hit by the end user? If that's the case, we should swap these to warning messages. We should only use `expect`/`unwrap` as a synonym for a `debug_assertion` to communicate to the developer that they did something wrong